pkg/ext/apiserver_authorization.go

```diff
@@ -0,0 +1,51 @@
+package ext
+
+import (
+	"context"
+
+	"github.com/rancher/steve/pkg/accesscontrol"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+)
+
+var _ authorizer.Authorizer = (*AccessSetAuthorizer)(nil)
+
+type AccessSetAuthorizer struct {
+	asl accesscontrol.AccessSetLookup
+}
+
+func NewAccessSetAuthorizer(asl accesscontrol.AccessSetLookup) *AccessSetAuthorizer {
+	return &AccessSetAuthorizer{
+		asl: asl,
+	}
+}
+
+func (a *AccessSetAuthorizer) Authorize(ctx context.Context, attrs authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	if !attrs.IsResourceRequest() {
+		// XXX: Implement
+		return authorizer.DecisionDeny, "", nil
+	}
+
+	verb := attrs.GetVerb()
+	namespace := attrs.GetNamespace()
+	name := attrs.GetName()
+	gr := schema.GroupResource{
+		Group:    attrs.GetAPIGroup(),
+		Resource: attrs.GetResource(),
+	}
+
+	accessSet := a.asl.AccessFor(attrs.GetUser())
+	if accessSet.Grants(verb, gr, namespace, name) {
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	return authorizer.DecisionDeny, "", nil
```

Comment:
The default behavior of returning `""` gives the following error:

```
testtypes.ext.cattle.io is forbidden: User \"unknown-user\" cannot list resource \"testtypes\" in API group \"ext.cattle.io\" at the cluster scope
```

If I return a non empty string (say: "because I say so"), then the message becomes:

```
testtypes.ext.cattle.io is forbidden: User \"unknown-user\" cannot list resource \"testtypes\" in API group \"ext.cattle.io\" at the cluster scope: because I say so
```

Your example above would duplicate the information, so I'd avoid adding it. We could provide _more_ information though, I'm just not sure what else would be useful. Any idea?

Reply:
If the default built-in provides enough context for users to know that verb/resource they don't have I think we can leave it empty. That being said, you should leave a comment to indicate that this is empty because the library provides sufficient context for the error.

Reply:
Addressed in b766a169e225934e6fd6dacdb3f66eea7337470e.